plots/plot_correlation_Npoints.py

```diff
@@ -0,0 +1,198 @@
+"""Plot a correlation plot showing the correlation between the number of data points in the training set and the metric chosen. The metric chosen can be MAE, RMSE, R2 or MAPE. The plot is saved in the folder plots/correlation_Npoints_vs_metric."""
+import os
+import logging
+import datetime
+
+import numpy as np
+
+import seaborn as sns
+import matplotlib.pyplot as plt
+import matplotlib.patheffects as path_effects
+from typing import List, Tuple, Optional, Callable, TypeVar, Generator, Union
+import pandas as pd
+
+Figure = TypeVar("Figure")
+from predictions.Prediction import PredictionData
+from plots.color_map_by_method import __color_map_by_method_dict
+from methods.plot_correlation_Npoints import (
+    plot_correlation_Npoints_vs_metric as method_report,
+)
+
+from data.report import Report
+
+
+def __get_dataset_name(results_dataframe: pd.DataFrame) -> str:
+    """Get the name of the dataset"""
+    return results_dataframe["dataset"].unique()[0]
+
+
+def __get_title(
+    results_dataframe: pd.DataFrame, chosen_metric: str, group_name: str
+) -> str:
+    """Get the title of the plot"""
+    if group_name is None:
+        return f"Correlation plot {chosen_metric} for {__get_dataset_name(results_dataframe)} for min/max of {results_dataframe['method'].unique().size} predictive methods on {results_dataframe['subset_row'].unique().size} datasets"
+    else:
+        return f"Correlation plot {chosen_metric} for {group_name} for min/max of {results_dataframe['method'].unique().size} predictive methods"
+
+
+def __plot_correlation_Npoints_vs_metric(
+    training_data: List[pd.DataFrame],
+    results_dataframe: pd.DataFrame,
+    chosen_metric: str = "MAE",
+    group_name: Optional[str] = None,
+) -> Figure:
+    """Plot correlation N data points in training set vs metric."""
+    figure, axis = plt.subplots(figsize=(12, 7))
+    axis.set_title(__get_title(results_dataframe, chosen_metric, group_name))
+    # condensed distance matrix must contain only finite values
+    results_dataframe[chosen_metric] = results_dataframe[chosen_metric].replace(
+        [np.inf, -np.inf], np.nan
+    )
+
+    pivoted_dataframe = results_dataframe.pivot(
+        columns="method", index="subset_row", values=chosen_metric
+    ).reset_index()
+    # condensed distance matrix must contain only finite values
+    pivoted_dataframe = pivoted_dataframe.replace([np.inf, -np.inf], np.nan)
+
+    # deal with the case where there are no results for a method
+    pivoted_dataframe = pivoted_dataframe.fillna(0)
+    data_lenght = []
+    for data in training_data:
+        data_lenght.append(len(data))
+    pivoted_dataframe_Ndata = pivoted_dataframe.copy()
+    pivoted_dataframe_Ndata["Ndata"] = data_lenght
+
+    if chosen_metric == "R2":
+        # create more ticks on the x axis
```

Comment:
Maybe get ChatGPT to help you restructure some of this, it seems slightly repetitive code.